### PR TITLE
Split RaftLogReader into CommittedReader and UncommittedReader

### LINKED ISSUE #6308: Remove JournalReader.MODE usage from RaftLog
**Description**

As part of #6147 - further decouple the `RaftLog` and `RaftLogReader` from the old journal by removing the usage of `JournalReader.MODE`.

1. Extract new interface `RaftLogReader`, which should wrap a `JournalReader` and deserialize the data of the `JournalRecord` into a `RaftLogEntry`
1. Create a new implementation `UncommittedRaftReader` which should read all entries unbounded.
1. Create a new implementation `CommittedRaftReader` which should read only committed entries.
1. Adapt `RaftLog` API to use these where applicable, and adjust where they are being called

Blocked by #6307

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

This PR splits the RaftLogReader into two - RaftLogCommittedReader, RaftLogUncommittedReader. The initial commits are refactoring necessary for the main change. 
In addition there is also a minor refactoring to make Sonarlint check passes. In this change, the reader in AtomixRecordEntrySupplier is changed to a volatile reader instead of a long living one. 

## Related issues

closes #6308 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLog.java [hunk 1]
```diff
@@ -62,7 +61,7 @@ public static RaftLogBuilder builder() {
    * @return the reader
    */
   public RaftLogReader openReader() {
```
Comment (pihme) [inline atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLog.java:63]: :thought_balloon: Here I am surprised by the method names. I would expect a pair of `openComittedReader` and `openUncommittedReader` But instead the commit reader has a special name, but the uncommitted doesn't have one.

I find this confusing. To me, using the uncommitted reader seems more dangerous, so I would expect to be warned by the method name.

File: atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLogUncommittedReader.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+import io.atomix.raft.storage.serializer.RaftEntrySerializer;
+import io.camunda.zeebe.journal.JournalReader;
+import io.camunda.zeebe.journal.JournalRecord;
+import java.util.NoSuchElementException;
+
+/**
+ * Raft log reader that reads both committed and uncommitted entries. This reader is supposed to be
+ * only used internally in raft. The reader is not thread safe with a concurrent writer.
+ */
+public class RaftLogUncommittedReader implements RaftLogReader {
```
Comment (pihme) [inline atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLogUncommittedReader.java:30]: :thought_balloon: I feel the name might be misleading. I might read this and think it reads only the uncommitted entries.

Maybe we can find a better name. Something like `RaftEagerReader` or `RaftLenientReader`, but as you can see, no obvious name comes to my mind, either

File: atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLogReader.java [hunk 3]
```diff
@@ -16,100 +15,49 @@
...
+  /**
+   * Seek to a record with the highest ASQN less than or equal to the given asqn, with the
+   * restriction that it seeks to a record with index less than or equal to the given
+   * indexUpperBound.
+   *
+   * @param asqn the asqn to seek to
+   * @param indexUpperBound the index until which it seeks
+   * @return the index of the record that will be returned by {@link #next()}
+   */
+  long seekToAsqn(final long asqn, final long indexUpperBound);
```
Comment (pihme) [inline atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLogReader.java:59]: :wrench: This method could be dropped. It is only called by `RaftCommittedReader`. Unless we anticipate a use for it in the future

File: atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLogUncommittedReader.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+import io.atomix.raft.storage.log.entry.RaftLogEntry;
+import io.atomix.raft.storage.serializer.RaftEntrySBESerializer;
+import io.atomix.raft.storage.serializer.RaftEntrySerializer;
+import io.camunda.zeebe.journal.JournalReader;
+import io.camunda.zeebe.journal.JournalRecord;
+import java.util.NoSuchElementException;
+
+/**
+ * Raft log reader that reads both committed and uncommitted entries. This reader is supposed to be
```
Comment (pihme) [inline atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLogUncommittedReader.java:27]: :thought_balloon: I wonder whether there is a way to make it impossible to access this class from outside Raft. E.g. by introducing an interface for `RaftLog` which doesn't contain the `openReader()` method. Then internally, we could use the implementation class, but external parties only see the method to get a committed reader.

**Review by pihme — COMMENTED** [review]
Looks good.

I do have a couple of questions, and some comments. Would be great if you could address those.

File: atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLogUncommittedReader.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+ * only used internally in raft. The reader is not thread safe with a concurrent writer.
+ */
+public class RaftLogUncommittedReader implements RaftLogReader {
+  private final JournalReader journalReader;
+  private final RaftEntrySerializer serializer = new RaftEntrySBESerializer();
+
+  RaftLogUncommittedReader(final JournalReader journalReader) {
```
[the remark this answers is not in the document]
Comment (deepthidevaki) [reply atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLogUncommittedReader.java:34]: This reader has to have a journal reader. So it should either get a journal or journalReader as the parameter. CommittedReader is a wrapper around the uncommittedReader. So it gets a RaftLog and creates an uncommittedReader.

File: atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLogUncommittedReader.java (hunk 2 above)
Comment (deepthidevaki) [reply atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLogUncommittedReader.java:30]: For me uncommittedReader sounds good. `Read Uncommitted` is a commonly used isolation level in databases which indicates it also allows to read uncommitted data, besides everything that is possible in stricter isolation levels. 

Alternate name I can think of is `UnrestrictedReader` and `CommittedReader`. What do you think?

File: atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLogUncommittedReader.java (hunk 4 above)
Comment (deepthidevaki) [reply atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLogUncommittedReader.java:27]: The API exposed via `RaftPartitionServer` only can open a committedReader. This is what external services use. Is that enough?

File: atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLogUncommittedReader.java (hunk 2 above)
Comment (pihme) [reply atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLogUncommittedReader.java:30]: I like the analogy with databases better.

**Review by pihme — APPROVED** [review]
thanks for answering my questions. 

lgtm :rocket:

**Comment by deepthidevaki** [comment]
bors retry

**Pull request merged**